### ISSUE TITLE
please post clarification (clearing cache clears ETags)

### ISSUE THREAD

**Issue by aeffioccau** [lucb1e/cookielesscookies#18]
there are erroneous articles being put up based upon your demo (e.g. https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/) which are claiming that browsers aren't deleting ETags when clearing the cache. They have reached this conclusion by a combination of the implication that the PoC is based upon ETag and/or cache itself, which is not true, as it's based off of IP address, and switching back and forth between IPs shows this. Please post a clarification of the fact that clearing the cache (in Firefox, at least) does in fact remove the ETag, and that the PoC still shows stored data is based upon re-populating the ETag based upon IP.

**Comment by claustromaniac** [comment]
How it is based off of IP address?

I'm not a web developer, so my understanding of the subject is limited, and the only part of the source code where I see the IP and user-agent being relevant is the single line where it generates the unique ETag for that client (index.php line 15):

`$etag = substr(sha1($secret . sha1($_SERVER["REMOTE_ADDR"]) . sha1($_SERVER["HTTP_USER_AGENT"])), 0, 18);`

Changing the IP after caching the tracker.jpg file indeed gives an unexpected result, but using the built-in Developer Tools in Firefox (don't know about other browsers), in the Network tab, you can see that the tracker.jpg file still has the same ETag even after you change the IP and reload the page.

So, I still don't get why this happens. I'm more inclined to think that something else is going on, and that this PoC does not rely on the client's IP staying the same by design.

Would you (or anyone else in the know) mind explaining what's going on?

**Comment by aeffioccau** [comment]
@claustromaniac 

The proof of concept is intended to demonstrate that ETag can be used as a permanent identifier, and it is true that it can be (so long as user doesn't clear cache, either manually or automatically), however, since each user's ETag is generated based upon a repeatable hash of a combination of their IP address and user-agent, the proof of concept works even if the cache is cleared and the request to tracker.jpg doesn't include the ETag. I will explain why momentarily ...

First, let me point out that, while not ideal, the method of generating the ETag does NOT invalidate the claim about how ETag can be used as a permanent identifier. It does however cause a ton of confusion, such as users and tech authors coming to believe that the ETags aren't getting deleted when their browser cache is cleared. Ways to address this could include:
- generating the ETag randomly, so that the PoC is based purely upon ETag, rather than using the IP (along with user agent) as the true identifier; or
- if for some reason there's objection to generating the ETag randomly, the PoC could instead detect when a request to tracker.jpg **doesn't** include an ETag, while the hash of the user's IP and user-agent for that connection matches an existing session, and invalidate that session; or
- at the very least, as requested in this bug, the author of the PoC could post a clarification explaining that clearing the cache really does remove the ETag, and the reason the demo still shows the stored text, even if the user cleared the ETag, is that the demo automatically regenerates the exact same ETag based upon the user's IP and user-agent.

Anyway, to explain to you exactly what's happening ...

Line 15 of index.php is as follows:
`$etag = substr(sha1($secret . sha1($_SERVER["REMOTE_ADDR"]) . sha1($_SERVER["HTTP_USER_AGENT"])), 0, 18);`

This is a programmatic formula which generates the same ETag given the same IP and user-agent. In short, the formula takes 3 strings (IP address (`$_SERVER["REMOTE_ADDR"]`), user-agent (`$_SERVER["HTTP_USER_AGENT"]`), and the "`$secret`" salt specified in config.php ("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")), runs them through the SHA-1 algorithm (an algorithm that, given the same strings, will always output the same result), and then keeps the first 18 character substring.

So, even if the cache is cleared, and the browser doesn't send the ETag when first requesting tracker.jpg after clearing the cache, tracker.jpg generates the same ETag automatically, tells the browser to store that value, and when the page is automatically reloaded (as it's explained that it does on the page itself), tracker.jpg is thus requested a second time, and the second request includes the ETag, and the page "thinks" it remembers you from cache, but (for users who cleared their cache), it's actually remembering you by IP address and user agent.

You can confirm all of this yourself using Firefox's Developer Tools Network pane. I'll quote an outline about how to reproduce this from https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/#comment-4306403 ...

> How to reproduce:
On your regular home ISP, clear your cache, then open the developer tools Network panel and go to the URL noted in prior paragraph [https://lucb1e.com/rp/cookielesscookies/tracker.jpg]. Click on the row for the tracker image and a pane will open on right. Go to the request headers and you'll see there is no If-None-Match header specifying the ETag. Then look at the response headers, and you'll see the same ETag as before. NEXT, close the page and clear your cache. Then use your mobile phone hotspot to connect your computer over cellular data. Repeat the same steps as above, and you'll see a different Etag. Then repeats the steps again on your home ISP and you'll get the original ETag. Then switch back to cellular and repeat and you'll see the cellular ETag. You can do this over and over and get the same two ETags, based upon your two different connections.

More simply, after closing the page and clearing the cache, if you open the Network Pane of Developer Tools and load http://lucb1e.com/rp/cookielesscookies/ , the second row will be a request for tracker.jpg -- click it, and in the right pane, scroll to "Request headers" you'll see there is no "If-None-Match" header (this is the header that tells the server what ETag you have stored). You will however see that the "Response headers" listed above specify the ETag, which is automatically generated by IP address and user-agent. The next request for tracker.jpg includes the "If-None-Match" header with the ETag which it was just told of. You can see this by reloading the page ... after reloading, click on the row for tracker.jpg, scroll down to the "Request headers", and you'll see the second request includes the "If-None-Match" header with the ETag.

In short: the PoC sets the ETag, which is stored until browser cache is cleared. When user goes to the page again after clearing cache, the ETag is (initially) regenerated based upon IP and user-agent, rather than retrieved from cache (because at that point it doesn't exist in cache) ... then the page automatically reloads (as stated on the page), and at that point the ETag the site stored a moment before is sent back, and it makes it look like you still had the ETag from much earlier whereas in reality you've only had it for about a second ... it just happens to be the same ETag from earlier because the same formula was used to generate it.

Clear as mud, right? This is why I've requested the author of the PoC post a clarification.

**Comment by claustromaniac** [comment]
We're talking about slightly different things. I'm not even thinking about clearing the cache at all. Change your IP _without_ clearing the cache and you will not get the result you would otherwise expect from the PoC. **This** is what puzzles me.

I think I must be missing some vital piece of information that prevents me from getting why that happens. You can see my interpretation of how this _should_ work (after looking at the source code) [here](https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/#comment-4306564) and [here](https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/#comment-4306660).
